### PR TITLE
[v3] rewrite general

### PR DESCRIPTION
**Pull request by palmtree5**
Rewrite of general.py for v3. Probably not completely done depending on what 26 wants to change in general (if anything) but it's a start

### CODE REVIEW

File: cogs/general/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from .general import General
+
+
+def setup(bot):
+    n = General(bot)
+    bot.add_listener(n.check_poll_votes, "on_message")
```
Comment (Rapptz) [inline cogs/general/__init__.py]: You can have `on_message` directly in the cog.

File: cogs/general/general.py [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+               }
+
+        if red_choice == player_choice:
+            outcome = None # Tie
+        else:
+            outcome = cond[(player_choice, red_choice)]
+
+        if outcome is True:
+            await ctx.send("{} You win {}!"
+                               "".format(red_choice.value, author.mention))
```
Comment (Rapptz) [inline cogs/general/general.py]: What is with the disgusting newline here?

File: cogs/general/general.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+    @commands.guild_only()
+    async def userinfo(self, ctx, *, user: discord.Member=None):
+        """Shows users's informations"""
+        author = ctx.message.author
+        guild = ctx.message.guild
+
+        if not user:
+            user = author
+
+        roles = [x.name for x in user.roles if x.name != "@everyone"]
```
Comment (Rapptz) [inline cogs/general/general.py]: `roles = [x.name for x in user.roles if not x.is_everyone()]`

File: cogs/general/general.py [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+
+        joined_at = self.fetch_joined_at(user, guild)
+        since_created = (ctx.message.timestamp - user.created_at).days
```
Comment (Rapptz) [inline cogs/general/general.py]: `Message.timestamp` no longer exists, use `Message.created_at`.

File: cogs/general/general.py [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+            msg = "(つ≧▽≦)つ" + name
+        elif intensity >= 10:
+            msg = "(づ￣ ³￣)づ{} ⊂(´・ω・｀⊂)".format(name)
+        await ctx.send(msg)
+
+    @commands.command()
+    @commands.guild_only()
+    async def userinfo(self, ctx, *, user: discord.Member=None):
+        """Shows users's informations"""
+        author = ctx.message.author
```
Comment (Rapptz) [inline cogs/general/general.py]: There is `Context.author`.

File: cogs/general/general.py [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+        guild = ctx.message.guild
```
Comment (Rapptz) [inline cogs/general/general.py]: There is `Context.guild`

File: cogs/general/general.py [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+        member_number = sorted(guild.members,
+                               key=lambda m: m.joined_at).index(user) + 1
+
+        created_on = "{}\n({} days ago)".format(user_created, since_created)
+        joined_on = "{}\n({} days ago)".format(user_joined, since_joined)
+
+        game = "Chilling in {} status".format(user.status)
+
+        if user.game is None:
+            pass
```
Comment (Rapptz) [inline cogs/general/general.py]: This statement is meaningless. Why are you even checking it?

File: cogs/general/general.py [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+        elif user.game.url is None:
```
Comment (Rapptz) [inline cogs/general/general.py]: I think you meant `if user.game and user.game.name`

File: cogs/general/general.py [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+            game = "Playing {}".format(user.game)
+        else:
+            game = "Streaming: [{}]({})".format(user.game, user.game.url)
+
+        if roles:
+            roles = sorted(roles, key=[x.name for x in guild.role_hierarchy
```
Comment (Rapptz) [inline cogs/general/general.py]: The first role in the hierarchy is the everyone role. You can remove it by doing something as basic as `guild.role_hierarchy[1:]`. I also don't know what this is supposed to be doing.

If you want to sort the user's roles by hierarchy the only thing you have to do is `sorted(member.roles)`.

File: cogs/general/general.py [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+        data.add_field(name="Joined Discord on", value=created_on)
+        data.add_field(name="Joined this guild on", value=joined_on)
+        data.add_field(name="Roles", value=roles, inline=False)
+        data.set_footer(text="Member #{} | User ID:{}"
+                             "".format(member_number, user.id))
+
+        name = str(user)
+        name = " ~ ".join((name, user.nick)) if user.nick else name
+
+        if user.avatar_url:
```
Comment (Rapptz) [inline cogs/general/general.py]: `User.avatar_url` now returns the default avatar URL if one is not found. You might want to change this to `if user.avatar` instead.

File: cogs/general/general.py [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+    @commands.guild_only()
+    async def guildinfo(self, ctx):
+        """Shows guild's informations"""
+        guild = ctx.message.guild
+        online = len([m.status for m in guild.members
+                      if m.status == discord.Status.online or
+                      m.status == discord.Status.idle])
+        total_users = len(guild.members)
+        text_channels = len([x for x in guild.channels
+                             if x.type == discord.ChannelType.text])
```
Comment (Rapptz) [inline cogs/general/general.py]: `type` attribute no longer exists. You might have meant `isinstance`.

You can do e.g. `len(guild.text_channels)` to get the text channel count.

File: cogs/general/general.py [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+        voice_channels = len(guild.channels) - text_channels
```
Comment (Rapptz) [inline cogs/general/general.py]: `len(guild.voice_channels)` or `len(guild.channels) - len(guild.text_channels)` from earlier.

File: cogs/general/general.py [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+        data = discord.Embed(
+            description=created_at,
+            colour=discord.Colour(value=colour))
+        data.add_field(name="Region", value=str(guild.region))
+        data.add_field(name="Users", value="{}/{}".format(online, total_users))
+        data.add_field(name="Text Channels", value=text_channels)
+        data.add_field(name="Voice Channels", value=voice_channels)
+        data.add_field(name="Roles", value=len(guild.roles))
+        data.add_field(name="Owner", value=str(guild.owner))
+        data.set_footer(text="Guild ID: " + guild.id)
```
Comment (Rapptz) [inline cogs/general/general.py]: Can't add `int` and `str`.

File: cogs/general/general.py [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+        text_channels = len([x for x in guild.channels
+                             if x.type == discord.ChannelType.text])
+        voice_channels = len(guild.channels) - text_channels
+        passed = (ctx.message.timestamp - guild.created_at).days
+        created_at = ("Since {}. That's over {} days ago!"
+                      "".format(guild.created_at.strftime("%d %b %Y %H:%M"),
+                                passed))
+
+        colour = ''.join([choice('0123456789ABCDEF') for x in range(6)])
+        colour = int(colour, 16)
```
Comment (Rapptz) [inline cogs/general/general.py]: Just use `random.randint` here. e.g. `random.randint(0, 0xFFFFFF)`.

File: cogs/general/general.py [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+                pos = int(search_terms[-1]) - 1
+                search_terms = search_terms[:-1]
+            else:
+                pos = 0
+            if pos not in range(0, 11): # API only provides the
+                pos = 0                 # top 10 definitions
+        except ValueError:
+            pos = 0
+
+        search_terms = "+".join([encode(s) for s in search_terms])
```
Comment (Rapptz) [inline cogs/general/general.py]: Honestly this entire function is fucked.

`aiohttp` supports `params` as a keyword argument and will automatically do this for you. I don't even know what's going on here.

File: cogs/general/general.py [hunk 16]
```diff
@@ -0,0 +1,433 @@
...
+        url = "http://api.urbandictionary.com/v0/define?term=" + search_terms
+        try:
+            async with aiohttp.request("GET", url) as r:
```
Comment (Rapptz) [inline cogs/general/general.py]: This was removed in aiohttp v2.0. Create a session and use it instead.

File: cogs/general/general.py [hunk 17]
```diff
@@ -0,0 +1,433 @@
...
+                result = await r.json()
+            if result["list"]:
```
Comment (Rapptz) [inline cogs/general/general.py]: Store this in a temporary variable instead of calling `__getitem__` repeatedly.

File: cogs/general/general.py [hunk 18]
```diff
@@ -0,0 +1,433 @@
...
+
+        Usage example:
+        poll Is this a poll?;Yes;No;Maybe
+        poll stop"""
+        message = ctx.message
+        if len(text) == 1:
+            if text[0].lower() == "stop":
+                await self.endpoll(message)
+                return
+        if not self.getPollByChannel(message):
```
Comment (Rapptz) [inline cogs/general/general.py]: Missing PEP-8.

File: cogs/general/general.py [hunk 19]
```diff
@@ -0,0 +1,433 @@
...
+            self.valid = False
+            return None
+        else:
+            self.valid = True
+        self.already_voted = []
+        self.question = msg[0]
+        msg.remove(self.question)
+        self.answers = {}
+        i = 1
+        for answer in msg: # {id : {answer, votes}}
```
Comment (Rapptz) [inline cogs/general/general.py]: You probably meant `for i, answer in enumerate(msg, 1)`.

**Review by Rapptz — CHANGES_REQUESTED** [review]
This is no good.

File: cogs/general/general.py (hunk 18 above)
Comment (palmtree5) [reply cogs/general/general.py]: Not completely clear on this one

EDIT: figured out you were referring to the function name being camel case as opposed to all lower, separated by underscores

**Comment by palmtree5** [comment]
Changes done

File: cogs/general/general.py [hunk 20]
```diff
@@ -201,16 +214,14 @@ def __init__(self, bot):
...
+        if user.game and user.game.name and user.game.url:
             game = "Streaming: [{}]({})".format(user.game, user.game.url)
+        elif user.game and user.game.name:
+            game = "Playing {}".format(user.game)
+
 
         if roles:
-            roles = sorted(roles, key=[x.name for x in guild.role_hierarchy
-                                       if x.name != "@everyone"].index)
+            roles = sorted(user.roles)
```
Comment (Rapptz) [inline cogs/general/general.py]: You probably want to just move this line to the top a bit.

First initialise:

`roles = sorted(user.roles)[1:]`

Then check if it exists and transform it:

```py
if roles:
  roles = ', '.join(x.name for x in roles)
```

**Pull request merged**